src/Microsoft.OData.Core/Json/JsonWriterExtensions.cs

```diff
@@ -206,5 +205,42 @@ private static void WriteJsonValue(this IJsonWriter jsonWriter, object propertyV
                 }
             }
         }
+
+        /// <summary>
+        /// Writes the json value (primitive, IDictionary or IEnumerable) to the underlying json writer.
+        /// </summary>
+        /// <param name="jsonWriter">The <see cref="JsonWriter"/> to write to.</param>
+        /// <param name="propertyValue">value to write.</param>
+        internal static void WriteODataValue(this IJsonWriter jsonWriter, ODataValue odataValue)
+        {
+            if (odataValue == null)
+            {
+                jsonWriter.WriteValue((string)null);
+                return;
+            }
+
+            object objectValue = odataValue.FromODataValue();
+            if (EdmLibraryExtensions.IsPrimitiveType(objectValue.GetType()))
+            {
+                jsonWriter.WritePrimitiveValue(objectValue);
+                return;
+            }
+
+            ODataResourceValue resourceValue = odataValue as ODataResourceValue;

```

Comment:
>odataValue as ODataResourceValue; [](start = 47, length = 33)

how about ODataCollectionValue?

Reply:
CollectionValue is out of scope because then I would need to know the EdmType of the collection which doesn't make a lot of sense for collection values. 

---
In reply to: [320877192](https://github.com/OData/odata.net/pull/1523#discussion_r320877192) [](ancestors = 320877192)

Reply:
The code should still be prepared to handle the case, though, right? even if it returns a (not supported?) error? Currently, it looks like we don't check for an ODataCollectionValue, so we'll just silently ignore?  Is that the desired behavior?

---
In reply to: [320936971](https://github.com/OData/odata.net/pull/1523#discussion_r320936971) [](ancestors = 320936971,320877192)